Set application/x-ndjson content type on bulk requests.

jmcarp opened the pull request:
Resolves #609

HonzaKral commented on elasticsearch/connection/http_urllib3.py:
```diff
@@ -111,6 +111,8 @@ def perform_request(self, method, url, params=None, body=None, timeout=None, ign
                 method = method.encode('utf-8')
 
+            request_headers = dict(self.headers)
+            request_headers.update(headers or {})
```
It would be nice if this block only ran `if headers:` to avoid copying the `self.headers` dict every time when it is only needed for the bulk-like bodies

HonzaKral commented on elasticsearch/client/__init__.py:
```diff
@@ -1164,7 +1164,8 @@ def bulk(self, body, index=None, doc_type=None, params=None):
         return self.transport.perform_request('POST', _make_path(index,
-            doc_type, '_bulk'), params=params, body=self._bulk_body(body))
+            doc_type, '_bulk'), params=params, body=self._bulk_body(body),
+            headers={'Content-Type': 'application/x-ndjson'})
```
we use lowercase for all headers to avoid conflicts - see https://github.com/elastic/elasticsearch-py/blob/master/elasticsearch/connection/http_urllib3.py#L59-L62

A test is needed to make sure the headers are merged properly

HonzaKral replied:
There are other APIs using the `_bulk_body` like `msearch` and `mpercolate`

HonzaKral commented on elasticsearch/transport.py:
```diff
@@ -269,6 +269,8 @@ def perform_request(self, method, url, params=None, body=None):
         :arg method: HTTP method to use
         :arg url: absolute url (without host) to target
+        :arg headers: dictionary of headers, will be handed over to the
+            underlying :class:`~elasticsearch.Connection` class for serialization
```
no serialization happen to the headers, we need to set expectations correctly

davidt99 commented:
@jmcarp Any plan to fix the review comments? Most of the driver requests (if not all) doesn't state content type, resulting warning logs in elastic 5.5.
The header support you wrote will be allow adding content type to all the request, thus fixing the warning.

HonzaKral commented:
@davidt99 All requests are anotated with content-type accepted by elasticsearch since d862c8a (released in `5.2`). You can also set headers globally by specifying `headers={'custom': 'header'}` when instantiating `Elasticsearch`.

davidt99 commented:
@HonzaKral Oh, I was using 5.1 and looked for similar issue. Thanks.

jmcarp commented:
I cleaned this up based on your suggestions @HonzaKral; let me know if this needs more revisions. I'm not sure why the tests are failing, but it looks like they're failing in the same way as `master`.

HonzaKral commented on elasticsearch/client/__init__.py:
```diff
@@ -1159,7 +1160,158 @@ def msearch(self, body, index=None, doc_type=None, params=None):
+            if param in SKIP_IN_PATH:
+                raise ValueError("Empty value passed for a required argument.")
+        return self.transport.perform_request('GET', _make_path(index,
+            doc_type, id, '_percolate', 'count'), params=params, body=body)
```
several APIs have been removed from `master` as they have been removed in elasticsearch `6.0`. Either please remove them from the patch or base the patch on `5.x`

HonzaKral commented on elasticsearch/client/__init__.py:
```diff
@@ -1387,4 +1540,3 @@ def field_caps(self, index=None, body=None, params=None):
         """
         return self.transport.perform_request('GET', _make_path(index,
             '_field_caps'), params=params, body=body)
-
```
please configure your editor not to make changes like these. It makes the PR hard to review when they are unrelated stylistic changes in unrelated places.

HonzaKral commented on elasticsearch/connection/base.py:
```diff
@@ -123,5 +123,3 @@ def _raise_error(self, status_code, raw_data):
 
         raise HTTP_EXCEPTIONS.get(status_code, TransportError)(status_code, error_message, additional_info)
-
-
```
please configure your editor not to make changes like these. It makes the PR hard to review when they are unrelated stylistic changes in unrelated places.

HonzaKral commented on elasticsearch/transport.py:
```diff
@@ -292,7 +294,7 @@ def perform_request(self, method, url, params=None, body=None):
         if body is not None:
             try:
-                body = body.encode('utf-8', 'surrogatepass')            
+                body = body.encode('utf-8', 'surrogatepass')
```
please configure your editor not to make changes like these. It makes the PR hard to review when they are unrelated stylistic changes in unrelated places.

HonzaKral commented on test_elasticsearch/test_server/test_helpers.py:
```diff
@@ -16,7 +16,7 @@ def bulk(self, *args, **kwargs):
         if self._called in self._fail_at:
             raise self._fail_with
-        return  self.client.bulk(*args, **kwargs)
+        return self.client.bulk(*args, **kwargs)
```
please configure your editor not to make changes like these. It makes the PR hard to review when they are unrelated stylistic changes in unrelated places.

jmcarp commented:
I rebased on master and dropped the deprecated methods, and reverted whitespace changes. This is ready for another look when you have time @HonzaKral.

fxdgear commented:
This LGTM. Can you please rebase off master and push again? I had to change the version of ES that travis uses.

jmcarp commented:
Done @fxdgear.

fxdgear commented:
Great thanks @jmcarp
